Missing type resolution for reference closure parameter?

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#[lang = "fn_once"]
trait FnOnce<T> {}

fn main() {
    let lambda = |&c| c != b'9';
}
```

I expected to see this happen: removing the lang item and feeding the lambda into `rustc`, we get this outptu

```rust
error[E0282]: type annotations needed for `&_`
 --> test.rs:3:19
  |
3 |     let lambda = |&c| c != b'9';
  |                   ^^ consider giving this closure parameter the explicit type `&_`, with the type parameters specified

error: aborting due to previous error
```

Instead, this happened: 
```rust
test.rs:6:19: error: expected T?, found reference
    6 |     let lambda = |&c| c != b'9';
      |                   ^
crab1: internal compiler error: in get_tyty, at rust/typecheck/rust-tyty-util.cc:40
0x21e2bb4 Rust::TyTy::TyVar::get_tyty() const
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-tyty-util.cc:40
0x21c386b Rust::TyTy::ReferenceType::get_base() const
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-tyty.cc:3042
0x2229ef4 Rust::Resolver::TypeCheckPattern::visit(Rust::HIR::ReferencePattern&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-pattern.cc:371
0x20c69ed Rust::HIR::ReferencePattern::accept_vis(Rust::HIR::HIRPatternVisitor&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/hir/tree/rust-hir.cc:4713
0x222828a Rust::Resolver::TypeCheckPattern::Resolve(Rust::HIR::Pattern*, Rust::TyTy::BaseType*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-pattern.cc:34
0x2234ffc Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::ClosureExpr&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:1490
0x20c6e9d Rust::HIR::ClosureExpr::accept_vis(Rust::HIR::HIRExpressionVisitor&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/hir/tree/rust-hir.cc:4863
0x222c3e6 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:41
0x22385be Rust::Resolver::TypeCheckStmt::visit(Rust::HIR::LetStmt&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-stmt.cc:84
0x20c6960 Rust::HIR::LetStmt::accept_vis(Rust::HIR::HIRStmtVisitor&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/hir/tree/rust-hir.cc:4695
0x22381a6 Rust::Resolver::TypeCheckStmt::Resolve(Rust::HIR::Stmt*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-stmt.cc:34
0x22307d6 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:554
0x20c72d0 Rust::HIR::BlockExpr::accept_vis(Rust::HIR::HIRExpressionVisitor&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/hir/tree/rust-hir.cc:4989
0x222c3e6 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:41
0x221cda9 Rust::Resolver::TypeCheckItem::visit(Rust::HIR::Function&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-item.cc:434
0x20c734d Rust::HIR::Function::accept_vis(Rust::HIR::HIRVisItemVisitor&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/hir/tree/rust-hir.cc:5001
0x2219658 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-item.cc:49
0x21b102d Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check.cc:37
0x200cba9 Rust::Session::compile_crate(char const*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/rust-session-manager.cc:620
0x200c257 Rust::Session::handle_input_files(int, char const**)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/rust-session-manager.cc:381
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
```

### Meta

#2138

--- a/gcc/rust/hir/tree/rust-hir-pattern.h
+++ b/gcc/rust/hir/tree/rust-hir-pattern.h
@@ -462,6 +462,8 @@ public:
 
   bool is_mut () const { return mut == Mutability::Mut; }
 
+  Mutability get_mutability () const { return mut; }
+
   void accept_vis (HIRFullVisitor &vis) override;
   void accept_vis (HIRPatternVisitor &vis) override;
 

--- a/gcc/rust/typecheck/rust-hir-type-check-expr.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-expr.cc
@@ -1471,25 +1471,20 @@ TypeCheckExpr::visit (HIR::ClosureExpr &expr)
   std::vector<TyTy::TyVar> parameter_types;
   for (auto &p : expr.get_params ())
     {
+      TyTy::BaseType *param_tyty = nullptr;
       if (p.has_type_given ())
 	{
-	  TyTy::BaseType *param_tyty
-	    = TypeCheckType::Resolve (p.get_type ().get ());
-	  TyTy::TyVar param_ty (param_tyty->get_ref ());
-	  parameter_types.push_back (param_ty);
-
-	  TypeCheckPattern::Resolve (p.get_pattern ().get (),
-				     param_ty.get_tyty ());
+	  param_tyty = TypeCheckType::Resolve (p.get_type ().get ());
 	}
       else
 	{
-	  TyTy::TyVar param_ty
-	    = TyTy::TyVar::get_implicit_infer_var (p.get_locus ());
-	  parameter_types.push_back (param_ty);
-
-	  TypeCheckPattern::Resolve (p.get_pattern ().get (),
-				     param_ty.get_tyty ());
+	  param_tyty = ClosureParamInfer::Resolve (p.get_pattern ().get ());
 	}
+
+      TyTy::TyVar param_ty (param_tyty->get_ref ());
+      parameter_types.push_back (param_ty);
+
+      TypeCheckPattern::Resolve (p.get_pattern ().get (), param_ty.get_tyty ());
     }
 
   // we generate an implicit hirid for the closure args

--- a/gcc/rust/typecheck/rust-hir-type-check-pattern.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-pattern.cc
@@ -447,5 +447,116 @@ TypeCheckPattern::visit (HIR::AltPattern &pattern)
 		 "type checking alternate patterns not supported");
 }
 
+TyTy::BaseType *
+ClosureParamInfer::Resolve (HIR::Pattern *pattern)
+{
+  ClosureParamInfer resolver;
+  pattern->accept_vis (resolver);
+
+  if (resolver.infered->get_kind () != TyTy::TypeKind::ERROR)
+    {
+      resolver.context->insert_implicit_type (resolver.infered);
+      resolver.mappings->insert_location (resolver.infered->get_ref (),
+					  pattern->get_locus ());
+    }
+  return resolver.infered;
+}
+
+ClosureParamInfer::ClosureParamInfer ()
+  : TypeCheckBase (), infered (new TyTy::ErrorType (0))
+{}
+
+void
+ClosureParamInfer::visit (HIR::WildcardPattern &pattern)
+{
+  HirId id = pattern.get_pattern_mappings ().get_hirid ();
+  infered = new TyTy::InferType (id, TyTy::InferType::InferTypeKind::GENERAL,
+				 TyTy::InferType::TypeHint::Default (),
+				 pattern.get_locus ());
+}
+
+void
+ClosureParamInfer::visit (HIR::IdentifierPattern &pattern)
+{
+  HirId id = pattern.get_pattern_mappings ().get_hirid ();
+  infered = new TyTy::InferType (id, TyTy::InferType::InferTypeKind::GENERAL,
+				 TyTy::InferType::TypeHint::Default (),
+				 pattern.get_locus ());
+}
+
+void
+ClosureParamInfer::visit (HIR::ReferencePattern &pattern)
+{
+  TyTy::BaseType *element
+    = ClosureParamInfer::Resolve (pattern.get_referenced_pattern ().get ());
+
+  HirId id = pattern.get_pattern_mappings ().get_hirid ();
+  infered = new TyTy::ReferenceType (id, TyTy::TyVar (element->get_ref ()),
+				     pattern.get_mutability ());
+}
+
+void
+ClosureParamInfer::visit (HIR::PathInExpression &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::StructPattern &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::TupleStructPattern &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::TuplePattern &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::LiteralPattern &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::RangePattern &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::QualifiedPathInExpression &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::SlicePattern &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
+void
+ClosureParamInfer::visit (HIR::AltPattern &pattern)
+{
+  rust_sorry_at (pattern.get_locus (),
+		 "unable to infer this kind of parameter pattern");
+}
+
 } // namespace Resolver
 } // namespace Rust

--- a/gcc/rust/typecheck/rust-hir-type-check-pattern.h
+++ b/gcc/rust/typecheck/rust-hir-type-check-pattern.h
@@ -58,6 +58,30 @@ private:
   TyTy::BaseType *infered;
 };
 
+class ClosureParamInfer : private TypeCheckBase, private HIR::HIRPatternVisitor
+{
+public:
+  static TyTy::BaseType *Resolve (HIR::Pattern *pattern);
+
+  void visit (HIR::PathInExpression &pattern) override;
+  void visit (HIR::StructPattern &pattern) override;
+  void visit (HIR::TupleStructPattern &pattern) override;
+  void visit (HIR::WildcardPattern &pattern) override;
+  void visit (HIR::TuplePattern &pattern) override;
+  void visit (HIR::LiteralPattern &pattern) override;
+  void visit (HIR::RangePattern &pattern) override;
+  void visit (HIR::IdentifierPattern &pattern) override;
+  void visit (HIR::QualifiedPathInExpression &pattern) override;
+  void visit (HIR::ReferencePattern &pattern) override;
+  void visit (HIR::SlicePattern &pattern) override;
+  void visit (HIR::AltPattern &pattern) override;
+
+private:
+  ClosureParamInfer ();
+
+  TyTy::BaseType *infered;
+};
+
 } // namespace Resolver
 } // namespace Rust
 

--- a/gcc/testsuite/rust/compile/issue-2142.rs
+++ b/gcc/testsuite/rust/compile/issue-2142.rs
@@ -1,0 +1,14 @@
+#[lang = "fn_once"]
+pub trait FnOnce<Args> {
+    #[lang = "fn_once_output"]
+    type Output;
+
+    extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
+}
+
+fn main() {
+    let lambda = |&c| c != b'9';
+
+    let a = b'1';
+    lambda(&a);
+}

